ArmPkg/Drivers/SmmuDxe/SmmuV3Util.c

```diff
@@ -0,0 +1,541 @@
+  if (ConsumerIndex == TotalQueueEntries) {
+    ConsumerIndex = 0;
+    ConsumerWrap  = ConsumerWrap ^ WrapMask;
+  }
+
+  Consumer.ReadIndex = ConsumerIndex | ConsumerWrap;
+
+  ArmDataSynchronizationBarrier ();
+
+  SmmuV3WriteRegister32 (SmmuInfo->SmmuBase + 0x10000, SMMU_EVENTQ_CONS, Consumer.AsUINT32);
+
+End:
+  return EFI_SUCCESS;
+}
+
+VOID
+EFIAPI
+SmmuV3PrintErrors (
+  IN SMMU_INFO  *SmmuInfo
+  )
+{
+  SMMUV3_GERROR        GError;
+  SMMUV3_FAULT_RECORD  FaultRecord = { 0 };
+
+  SmmuV3ConsumeEventQueueForErrors (SmmuInfo, &FaultRecord);
+  DEBUG ((DEBUG_INFO, "FaultRecord:\n"));
+  for (UINTN i = 0; i < sizeof (FaultRecord.Fault) / sizeof (FaultRecord.Fault[0]); i++) {
+    DEBUG ((DEBUG_INFO, "0x%llx\n", FaultRecord.Fault[i]));
+  }
+
+  GError.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_GERROR);
+  DEBUG ((DEBUG_INFO, "GError: 0x%lx\n", GError.AsUINT32));
+}
+
+STATIC
+VOID
+EFIAPI
+SmmuV3WriteCommands (
+  IN SMMU_INFO           *SmmuInfo,
+  IN UINT64              StartingIndex,
+  IN UINT32              CommandCount,
+  IN SMMUV3_CMD_GENERIC  *Commands
+  )
+{
+  UINT32              Index;
+  UINT32              ProducerIndex;
+  UINT32              QueueMask;
+  UINT32              WrapMask;
+  SMMUV3_CMD_GENERIC  *CommandQueue;
+
+  WrapMask     = (1UL << SmmuInfo->CommandQueueLog2Size);
+  QueueMask    = WrapMask - 1;
+  CommandQueue = (SMMUV3_CMD_GENERIC *)SmmuInfo->CommandQueue;
+  for (Index = 0; Index < CommandCount; Index += 1) {
+    ProducerIndex               = (UINT32)((StartingIndex + Index) & QueueMask);
+    CommandQueue[ProducerIndex] = Commands[Index];
+  }
+}
+
+EFI_STATUS
+EFIAPI
+SmmuV3SendCommand (
+  IN SMMU_INFO           *SmmuInfo,
+  IN SMMUV3_CMD_GENERIC  *Command
+  )
+{
+  UINT32            QueueMask;
+  UINT32            WrapMask;
+  UINT32            TotalQueueEntries;
+  UINT32            NewProducerIndex;
+  UINT32            ProducerIndex;
+  UINT32            ConsumerIndex;
+  UINT32            ProducerWrap;
+  UINT32            ConsumerWrap;
+  SMMUV3_CMDQ_PROD  Producer;
+  SMMUV3_CMDQ_CONS  Consumer;
+  UINT8             Count;
+
+  // Set 1ms timeout value.
+  Count = 10;
+
+  Producer.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_CMDQ_PROD);
+  Consumer.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_CMDQ_CONS);
+
+  TotalQueueEntries = SMMUV3_COUNT_FROM_LOG2 (SmmuInfo->CommandQueueLog2Size);
+  WrapMask          = TotalQueueEntries;
+  QueueMask         = WrapMask - 1;
+  ProducerWrap      = Producer.WriteIndex & WrapMask;
+  ConsumerWrap      = Consumer.ReadIndex & WrapMask;
+
+  ProducerIndex = Producer.WriteIndex & QueueMask;
+  ConsumerIndex = Consumer.ReadIndex & QueueMask;
+
+  while (Count > 0 && SMMUV3_IS_QUEUE_FULL (
+                        ProducerIndex,
+                        ProducerWrap,
+                        ConsumerIndex,
+                        ConsumerWrap
+                        ) != FALSE)
+  {
+    MicroSecondDelay (100);

```

Comment:
is 100 microseconds a spec defined value?

Reply:
I use an arbitrary value here, unclear on what to use as a timeout value.